File: packages/helpers/src/obligationStatus.ts
// Copyright (c) 2025-2026 VATM ICPMS <sms@vatm.vn>.
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
// AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
// OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
// PERFORMANCE OF THIS SOFTWARE.

type Translator = (s: string) => string;

export type ObligationStatus = "NON_COMPLIANT" | "PARTIALLY_COMPLIANT" | "COMPLIANT";

export const obligationStatuses = [
  "NON_COMPLIANT",
  "PARTIALLY_COMPLIANT",
  "COMPLIANT",
] as const;

export const getObligationStatusVariant = (status: ObligationStatus) => {
  switch (status) {
    case "NON_COMPLIANT":
      return "danger" as const;
    case "PARTIALLY_COMPLIANT":
      return "warning" as const;
    case "COMPLIANT":
      return "success" as const;
    default:
      return "neutral" as const;
  }
};

export const getObligationStatusLabel = (status: ObligationStatus) => {
  switch (status) {
    case "NON_COMPLIANT":
      return "Non-compliant";
    case "PARTIALLY_COMPLIANT":
      return "Partially compliant";
    case "COMPLIANT":
      return "Compliant";
    default:
      return status;
  }
};

export function getObligationStatusOptions(__: Translator) {
  return obligationStatuses.map((status) => ({
    value: status,
    label: __({
      "NON_COMPLIANT": "Non-compliant",
      "PARTIALLY_COMPLIANT": "Partially compliant",
      "COMPLIANT": "Compliant",
    }[status]),
  }));
}

File: pkg/coredata/obligation_status.go
// Copyright (c) 2025-2026 VATM ICPMS <sms@vatm.vn>.
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
// AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
// OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
// PERFORMANCE OF THIS SOFTWARE.

package coredata

import (
	"encoding"
	"fmt"
)

type ObligationStatus string

const (
	ObligationStatusNonCompliant       ObligationStatus = "NON_COMPLIANT"
	ObligationStatusPartiallyCompliant ObligationStatus = "PARTIALLY_COMPLIANT"
	ObligationStatusCompliant          ObligationStatus = "COMPLIANT"
)

var (
	_ fmt.Stringer             = ObligationStatus("")
	_ encoding.TextMarshaler   = ObligationStatus("")
	_ encoding.TextUnmarshaler = (*ObligationStatus)(nil)
)

func ObligationStatuses() []ObligationStatus {
	return []ObligationStatus{
		ObligationStatusNonCompliant,
		ObligationStatusPartiallyCompliant,
		ObligationStatusCompliant,
	}
}

func (v ObligationStatus) IsValid() bool {
	switch v {
	case
		ObligationStatusNonCompliant,
		ObligationStatusPartiallyCompliant,
		ObligationStatusCompliant:
		return true
	}

	return false
}

func (v ObligationStatus) String() string {
	return string(v)
}

func (v ObligationStatus) MarshalText() ([]byte, error) {
	return []byte(v.String()), nil
}

func (v *ObligationStatus) UnmarshalText(text []byte) error {
	val := ObligationStatus(text)
	if !val.IsValid() {
		return fmt.Errorf("invalid ObligationStatus value: %q", string(text))
	}

	*v = val

	return nil
}
